### ISSUE TITLE
Question about serializable objects and blocking

### ISSUE THREAD

**Issue by D3SL** [shikokuchuo/mirai#1]
Two questions: Does Mirai have the same limitations as callr and future in that many things like DB connections are unserializable and therefore a new connection has to be established inside every async code block? And is mirai self-blocking if used in a Shiny app, or can it be used for true same-session non-blocking async code?

**Comment by shikokuchuo** [comment]
- Does Mirai have the same limitations as callr and future in that many things like DB connections are unserializable and therefore a new connection has to be established inside every async code block?

Objects are passed serialised to the new process, so any limitations with R’s native serialisation will be shared with mirai.

If you have a custom application, it is possible to implement yourself using {nanonext} if you were so minded. You would set up an RPC server with the open DB connection, and a client would make requests to it – there are even convenience functions `request()` and `reply()` in that package that incorporate the logic. {mirai} is really a proof of concept and uses only publicly-exported functions from {nanonext}.

- And is mirai self-blocking if used in a Shiny app, or can it be used for true same-session non-blocking async code?

`call_mirai()` is blocking – it will wait until the mirai is resolved. The logic being that the value would be called at the point where subsequent code depends on it, and hence the correct thing to do is wait. 

However, we intend the next version of ‘mirai’ to also feature a non-blocking call. It would then be plausible to use that inside something like `shiny::req()`. This functionality is already under test in the development version of {nanonext}.

Note that if you do not need the value of the mirai back, for example it is doing some I/O activity, you can send an `eval_mirai()` without calling it and this is completely non-blocking. However, you will also not capture any errors so only do this if you are prepared for the action to potentially fail.

I hope the above answers your questions?

**Comment by shikokuchuo** [comment]
The non-blocking call is now implemented in 328e977756071475c923811dbd2bde49aa7c320c for testing.

The dev version of mirai now requires the dev version of nanonext. To install both, set the following option:
 ```
options(repos = c(shikokuchuo = 'https://shikokuchuo.r-universe.dev',
                  CRAN = 'https://cloud.r-project.org'))
```
And then proceed with:  `install.packages("mirai")`

**Comment by shikokuchuo** [comment]
0d983b0393d7116e52460faa7dbccdc24600493e fully implements non-blocking. This is no longer marked experimental. 

Any further questions please feel free to re-open or open another issue.
